### DISCUSSION TITLE
Modeling badminton racket as an elastic rod using cable

### DISCUSSION THREAD

**Discussion by alex1yaremchuk** [google-deepmind/mujoco#1386] [General]
Hi,

I'm an amateur researcher and I'm trying to use MuJoCo for modeling a badminton racket (later with an arm).
I'm not a physicist but I'm a programmer.

I'm aiming at modeling a racket rotating around a wrist for the beginning. 
I want to compare speeds of a rigid stick and flexible rod ends and therefore measure the whipping effect. 
Later I wanted to model shoulder-elbow-wrist system and find optimal swing using reinforcement learning.

I'm looking for some help with understanding how to simulate an elastic rod of approximate badminton racket characteristics. 
I was successfull at rigging up a flexible stick on a hinge which can be rotated using an actuator. 

But when I try to play with it, f.i. make it thinner it becomes unstable. 
I have to use quite a small timestep to make it work. 

**Questions:** 
1) Can I have a general advice - am I doing right or my simulation can be made more robust? 
I need it to be quite effective as I suppose to use it for RL later - so it should calculate quickly.

2) Can I have an advice on bringing my rod closer to real badminton racket on physical characteristics?

Here is a notebook I'm experimenting with  (section Flexible rod based on cable)  - 
https://github.com/google-deepmind/mujoco#asking-questions

Here is a model from this notebook:

<details>
  <summary>minimal XML</summary>

```XML
<mujoco model="smash">
  <option integrator="RK4" timestep="0.00002"/>
  <extension>
    <plugin plugin="mujoco.elasticity.cable"/>
  </extension>
  <asset>
    <texture name="grid" type="2d" builtin="checker"
      width="300" height="400" />
    <material name="grid" texture="grid" texrepeat="8 8"
      reflectance=".2" />
  </asset>

  <compiler autolimits="true"/>

 <size memory="200M" />


  <worldbody>
    <light diffuse="10.5 10.5 10.5" name="top" pos="0 0 6" dir="0 0 -1"/>
<!--    <camera name="cam1" pos="0 -7 4" euler="80 0 0" /> -->
<!-- doesn't work   <camera name="cam1" pos="0 -7 4" axisangle="1 1 1 0 " /> -->
    <camera name="cam1" pos="0 -7 4" zaxis="0 -1 0.2"  />
    <!-- <geom size="10 10 .1" type="plane" material="grid" /> -->
    <geom size="10 10 .1" type="plane" rgba="0 .9 0 1" />

    <body name="shoulder_body" >
      <joint name="swing" type="hinge" axis="0 1 0" pos="0 0 3"/>
      <composite type="cable" curve="s" 
        count="20 1 1" size="2" offset="0 0 3" initial="none">
        <plugin plugin="mujoco.elasticity.cable">
          <!--Units are in Pa (SI)
          <config key="twist" value="1e7"/>
          <config key="bend" value="6e6"/>
          <config key="vmax" value="0.05"/> -->

          <config key="twist" value="1e9"/>
          <config key="bend" value="1e8"/>
          <config key="vmax" value="0.05"/>

        </plugin>
        <joint kind="main" damping=".015"/> 
        <geom type="capsule" size=".05 .5" rgba="1 0 0 1" 
          condim="1"/>
      </composite>
    </body>
  </worldbody>
<!--  <keyframe>
    <key name="swinging" qvel="20"/>
    <key name="steady" qvel="0"/>
  </keyframe> -->
  <actuator>
    <motor name="swing_motor" joint="swing" ctrlrange="-1000 1000"/>
  </actuator>
</mujoco>
```

</details>

Here is a video of simple simulation of applying force for a second and opposite force later.

https://github.com/google-deepmind/mujoco/assets/40666786/a0f9b46a-257b-4380-8ae3-235f16827bc8

All feedback is appreciated!

**Comment by Balint-H** [comment]
I was able to get the rod to behave in a similar way at 100 times larger timestep with the following attributes inside the joint definition of the composite body:
``` xml
<joint kind="main" damping="20" stiffness="2000" armature="0.01"/> 
```
If it's too wobbly, bump up the stiffness to 10000<.


I'm not sure how specifying stiffness there interacts with the `twist` and `bend` parameters of the cable though. I would have though the stiffness would be determined by them, but specifying both cable properties and stiffness seems to be necessary for a stable simulation. Armature in general helps stabilize exploding joints.

Also, the `implicit` integrator seems to be more stable than RK4 here.

**Reply by alex1yaremchuk** [reply]
Thank you very much! 

The main result - it's much better with "stiffness" specified as for stability. 
And I was able to reduce "bend" with about the same look.

my simulation really felt wobbly - I even doubted it's elasticity, it looked like a rope. 

I think I'll file a question into Issues asking to clarify these params meaning - stiffness of joint vs bend of a cable. 

Another question for me is the length of the composite. 
With the same camera setup for a rigid stick I have a capsule of length 1. 
And here I got 20 segment composite of capsules of size=".03 .5" 
but overall length is not count * size[1] (second param of sizeI) , 
Length has to do something with composite.size 
which I had to set to 2 to have it look like a capsule of length 1.

For now it looks like this, made it thinner, stiffer, about the same bending:
https://github.com/google-deepmind/mujoco/assets/40666786/d6397848-41cd-4cd8-8ac0-0ade89c3412f

**Reply by alex1yaremchuk** [reply]
An update to my little research - 

1) was able to make cable elastic rod stable enough

2) added speed sensor to the stick end (to both rigid capsule stick and the composite)

3) was able to reproduce the whipping effect.
(if it is made really wobbly then it can make even times bigger speed)

4) still unclear about "bend", "stiffness". 
Ideally I'd like to change elasticity from sosage like flexible stick to absolutely rigid one.
in this case I won't need simple capsule stick at all.

5) still no ML,  just exploring the model for now.

wobbly one (stiffness 10^3, bend 10^9) speed 240:

https://github.com/google-deepmind/mujoco/assets/40666786/12a50757-386f-4182-a442-f2463745c101

stiff one (stiffness 10^5, bend 10^9)  - speed 80

https://github.com/google-deepmind/mujoco/assets/40666786/13c34a63-b9da-4530-b173-fadb1d3bb243

**Reply by Balint-H** [reply]
Heya when you say absolutely rigid do you mean no bend at all? Because that is of course very easy.

Decreasing the number of segments will also likely decrease the wobbliness and make it more stable.

Also play a bit with the armature arguement, I added that to prevent exploding, but too much can add unneeded inertia to the rod, also making it wobbly.

**Reply by alex1yaremchuk** [reply]
Hey! thank you for your prompt reply, I'll give it a try.

Yeap, I thought of having a scale from absolutely rigid one to a flexible stick. 
I had in mind showing 
1) impact of whipping effect on the resulting speed (I'd like to base whipping effect on absolutely rigid one)
2) dependency on the motor power (player strength) and the optimal flexibility of the racket corresponding to it.

**Reply by alex1yaremchuk** [reply]
Yeah, it becomes unstable with lower armature..

**Reply by alex1yaremchuk** [reply]
As for number of segments - probably I don't need 20. 

But I'm not clear about that - initially I thought that the more parts  - the closer to real rod it will be. 
(I don't care about smooth rendering but only about adequate modelling)

What I'm expecting from the flexibility model is that with increasing flexibility the rod is going to be bend more in a bow as a whole. 
but what I see now - it behaves like a rope or a sausage. 

I have some doubts  - did I correctly choose cable, maybe I need a rope for such a lengthty object?

**Reply by Balint-H** [reply]
@quagla can you advise how the elasticity plugin could be used to achieve this kind of high stiffness bending?

**Comment by alex1yaremchuk** [comment]
Another idea - may be Mujoco is not the right emulation framework for my needs as I don't need any contacts or some other complicated niceties. 
I just need a rod simulation (and later several joints arm simulation)

**Comment by quagla** [comment]
Discussed and solved in #1399.
